pkg/codegen/schema/schema.go

```diff
@@ -0,0 +1,898 @@
+	ArchiveType Type = primitiveType(archiveType)
+	// AssetType represents the set of Pulumi Asset values.
+	AssetType Type = primitiveType(assetType)
+	// AnyType represents the complete set of values.
+	AnyType Type = primitiveType(anyType)
+)
+
+// MapType represents maps from strings to particular element types.
+type MapType struct {
+	// ElementType is the element type of the map.
+	ElementType Type
+}
+
+func (t *MapType) String() string {
+	return fmt.Sprintf("Map<%v>", t.ElementType)
+}
+
+func (*MapType) isType() {}
+
+// ArrayType represents arrays of particular element types.
+type ArrayType struct {
+	// ElementType is the element type of the array.
+	ElementType Type
+}
+
+func (t *ArrayType) String() string {
+	return fmt.Sprintf("Array<%v>", t.ElementType)
+}
+
+func (*ArrayType) isType() {}
+
+// UnionType represents values that may be any one of a specified set of types.
+type UnionType struct {
+	// ElementTypes are the allowable types for the union type.
+	ElementTypes []Type
+}
+
+func (t *UnionType) String() string {
+	elements := make([]string, len(t.ElementTypes))
+	for i, e := range t.ElementTypes {
+		elements[i] = e.String()
+	}
+	return fmt.Sprintf("Union<%v>", strings.Join(elements, ", "))
+}
+
+func (*UnionType) isType() {}
+
+// ObjectType represents schematized maps from strings to particular types.
+type ObjectType struct {
+	// Token is the type's Pulumi type token.
+	Token string
+	// Comment is the description of the type, if any.
+	Comment string
+	// Properties is the list of the type's properties.
+	Properties []*Property
+	// Language specifies additional language-specific data about the object type.
+	Language map[string]json.RawMessage
+}
+
+func (t *ObjectType) String() string {
+	return t.Token
+}
+
+func (*ObjectType) isType() {}
+
+// TokenType represents an opaque type that is referred to only by its token. A TokenType may have an underlying type
+// that can be used in place of the token.
+type TokenType struct {
+	// Token is the type's Pulumi type token.
+	Token string
+	// Underlying type is the type's underlying type, if any.
+	UnderlyingType Type
+}
+
+func (t *TokenType) String() string {
+	return t.Token
+}
+
+func (*TokenType) isType() {}
+
+// DefaultValue describes a default value for a property.
+type DefaultValue struct {
+	// Value specifies a static default value, if any. This value must be representable in the Pulumi schema type
+	// system, and its type must be assignable to that of the property to which the default applies.
+	Value interface{}
+	// Environment specifies a set of environment variables to probe for a default value.
+	Environment []string
+	// Language specifies additional language-specific data about the default value.
+	Language map[string]json.RawMessage
+}
+
+// Property describes an object or resource property.
+type Property struct {
+	// Name is the name of the property.
+	Name string
+	// Comment is the description of the property, if any.
```

Comment:
no need for deprecation?

Reply:
IIRC we don't deprecate individual properties at the moment. Easy enough to add support now or later, though.